### ISSUE TITLE
Offline auth potentially dangerous?

### ISSUE THREAD

**Issue by michalwiglasz** [mergado/mergado-apps#29]
I cannot confirm that, but I have feeling that offline authorization can be dangerous and lead to data leak. Let's assume that I have app which talks to API using only offline-authorized tokens (without user_id). If I understand it correctly, I can take iframe URL from my eshop:
```
https://appcloud.mergado.com/apps/myapp/eshop/950
```

And try changing ID until I find an eshop which has this app enabled:
```
https://appcloud.mergado.com/apps/myapp/eshop/951  # nope
https://appcloud.mergado.com/apps/myapp/eshop/952  # nope
...
https://appcloud.mergado.com/apps/myapp/eshop/965  # win!
```

In this case, afaik there is no way to tell, that the current user has no access to eshop 965 (user id was not passed to API, login session is not shared between Mergado and App), yet the app can show confidental information. 

---

Am I correct?

**Comment by paveldedik** [comment]
With each access token, there is a subject_id (stored with the generated token), this can be either user_id, eshop_id or project_id. Offline tokens have special permission based on the type of the subject. If it is a project, the client is only permitted to access or modify that one project (can access its basic info, rules, queries, elements etc.). If the subject is an eshop, the client can access or modify only the eshop, its projects, rules belonging to one of the projects, etc. So by iterating over all eshop IDs, you would either reach an eshop the token is allowed to read, or your program would loop forever.

TL;DR offline and online tokens have different permissions. Offline token -> permissions are derived from a subject. Online token -> permissions are always derived from the authenticated user.

**Comment by michalwiglasz** [comment]
But if I navigate to different project, the app can request new access token with correct `subject_id`, right? Let's say I navigate to `https://appcloud.mergado.com/apps/myapp/eshop/965`, the app requests access token for `eshop_id = 965` and can leak eshop data even I do not have access rights to it.

**Comment by michalwiglasz** [comment]
What I'm trying to say is, that the app can effectively bypass user authorization. If only online auth were possible, this could not happen.

**Comment by paveldedik** [comment]
You'd have to have access rights to it (the user would have to enable the app beforehand for this subject), otherwise the authorization server wouldn't generate an access token for the client and returned 403 instead. The client is authorized with an ID and a secret key, the authorization server searches the DB and decides whether to generate an access token or not. See Refresh Token grant type in OAuth2 protocol: [RFC6749](https://tools.ietf.org/html/rfc6749#section-1.5)

**Comment by michalwiglasz** [comment]
If I open https://appcloud.mergado.com/apps/biddingrank.dev/eshop/2544/project/11724/policies/new

I can clearly see the project queries, although I do NOT have access to that project. Hm?

**Comment by paveldedik** [comment]
It seems that you have found a bug! If I understand correctly, you used a refresh token to obtain an access token for another eshop, right? Anyway, don't use refresh token in online mode (i.e. when the user is authenticated), that should solve the problem.

Thanks for reporting the issue (and standing up for it).

**Comment by michalwiglasz** [comment]
My app uses only offline tokens (easier to implement & it works so far). I never obtained online tokens.

**Comment by michalwiglasz** [comment]
Generally, following the docs, I implemented my app so it requests offline token (refresh token generated from subject id) when necessary. Shouldn't it be limited so I need online token to request an offline one?

**Comment by paveldedik** [comment]
Well, you shouldn't be using offline token for online access, that's not what they meant for (and now, when we know there is a bug, it's even dangerous and your app could leak confidental information)

**Comment by michalwiglasz** [comment]
Shouldn't this be mentioned somewhere in docs? Like in huge red letters with blinking frame?

**Comment by michalwiglasz** [comment]
Anyway, I think this should be solved on some other level than note in docs. Anyone can access the API from code not under your control - afaik API access is not limited to appcloud ip range, right?

**Comment by paveldedik** [comment]
We didn't know there was a bug. It escalated from "offline tokens are not supposed to be used in online mode" to "offline tokens are dangerous when used in online mode" :-/
